crates/fuel-core/src/service/genesis/importer/import_task.rs

```diff
@@ -27,144 +30,134 @@ use crate::{
 where
-    DbDesc: DatabaseDescription,
+    T: TableWithBlueprint,
 {
-    handler: Handler,
-    skip: usize,
-    groups: Groups,
-    db: GenesisDatabase<DbDesc>,
-    reporter: ProgressReporter,
-}
-
-pub trait ImportTable {
-    type TableInSnapshot: TableWithBlueprint;
-    type TableBeingWritten: TableWithBlueprint;
-    type DbDesc: DatabaseDescription;
+    fn on_chain(
+        &mut self,
+        _group: Cow<Vec<TableEntry<T>>>,
+        _tx: &mut StorageTransaction<&mut GenesisDatabase<OnChain>>,
+    ) -> anyhow::Result<()> {
+        Ok(())
+    }
 
-    fn process(
+    fn off_chain(
         &mut self,
-        group: Vec<TableEntry<Self::TableInSnapshot>>,
-        tx: &mut StorageTransaction<&mut GenesisDatabase<Self::DbDesc>>,
-    ) -> anyhow::Result<()>;
+        _group: Cow<Vec<TableEntry<T>>>,
+        _tx: &mut StorageTransaction<&mut GenesisDatabase<OffChain>>,
+    ) -> anyhow::Result<()> {
+        Ok(())
+    }
 }
 
-impl<Logic, GroupGenerator, DbDesc> ImportTask<Logic, GroupGenerator, DbDesc>
+pub fn import_entries<T>(
+    cancel_token: CancellationToken,
+    mut handler: impl ImportTable<T>,
+    groups: impl IntoIterator<Item = anyhow::Result<Vec<TableEntry<T>>>>,
+    mut on_chain_db: GenesisDatabase<OnChain>,
+    mut off_chain_db: GenesisDatabase<OffChain>,
+    reporter: ProgressReporter,
+) -> anyhow::Result<()>
 where
-    DbDesc: DatabaseDescription,
-    Logic: ImportTable<DbDesc = DbDesc>,
-    GenesisDatabase<DbDesc>: StorageInspect<GenesisMetadata<DbDesc>>,
+    T: TableWithBlueprint,
 {
-    pub fn new(
-        handler: Logic,
-        groups: GroupGenerator,
-        db: GenesisDatabase<DbDesc>,
-        reporter: ProgressReporter,
-    ) -> Self {
-        let progress_name =
-            migration_name::<Logic::TableInSnapshot, Logic::TableBeingWritten>();
-        let skip = match db.storage::<GenesisMetadata<DbDesc>>().get(&progress_name) {
-            Ok(Some(idx_last_handled)) => {
-                usize::saturating_add(idx_last_handled.into_owned(), 1)
-            }
-            _ => 0,
-        };
-
-        Self {
-            handler,
-            skip,
-            groups,
-            db,
-            reporter,
+    let on_chain_last_idx = on_chain_db
+        .storage::<GenesisMetadata<OnChain>>()
+        .get(T::column().name())?
+        .map(|x| x.into_owned());
+
+    let off_chain_last_idx = off_chain_db
+        .storage::<GenesisMetadata<OffChain>>()
+        .get(T::column().name())?
+        .map(|x| x.into_owned());
+
+    let num_groups_handled_by_all_dbs = on_chain_last_idx
+        .min(off_chain_last_idx)
+        .map(|x| x.saturating_add(1))
+        .unwrap_or(0);
+
+    for (index, group) in groups
+        .into_iter()
+        .enumerate()
+        .skip(num_groups_handled_by_all_dbs)
+    {
+        if cancel_token.is_cancelled() {
+            bail!("Import cancelled");
+        }
+        let group = group?;
+
+        if Some(index) > on_chain_last_idx {
```

Comment:
If no batch is handled then the database has no entry for that table. i.e. a select would return 0 rows. Here given as `None` from the db. 

When a batch is handled it's index (parquet group number) is recorded.

So an entry of `0` in the database means the first batch is handled successfully.

So here the resume logic is identical to the comparison of options. If there is no last recorded idx for the on chain database then no batch was imported and so proceed to import.

If there is an entry in the db check that this index is greater so you don't reimport the same batch.

But `unwrap_or(0)` would cause problems for the first batch because of the difference in behavior you mentioned.

I agree the options comparison is unclear if you're coming across it for the first time. Maybe the second time as well :D. 

The alternative would be to unpack the behavior:
```rust
let comparison_result = if let Some(chain_idx) = on_chain_last_idx {
    index > chain_idx
} else {
    true
};
```